azure-mssql-fog-run-failover.yml

```diff
@@ -41,6 +41,7 @@ provision:
     type: object
     details: 'JSON hash of server pair name to set of server names. { "name": { "primary":{"server_name":"...", "resource_group":..."}, "secondary":{"server_name":"...", "resource_group":..."}, ...}'
     required: true
+    default: ${config("service.csb-azure-mssql-fog-run-failover.provision.defaults.server_pairs")}
```

Comment:
thoughts: the default values are automatically propagated to the context and then transformed into Terraform variables.
The change added implies being more explicit and I like it because it helps to understand the operations performed under the hood. This modification would need a team discussion to see if we prefer to be more intentionally explicit and, from now on, define that strategy.

On the other hand, I would love to know what happens if there are no server_pairs defined. Will it break?


See the line where we merge the defaults for the services and add the values of those properties to the final context: [xxx](https://github.com/cloudfoundry/cloud-service-broker/blob/f9d0df3465d3cb4d6cba98796998a178fc184682/pkg/broker/service_definition.go#L406).

Something similar happens when provisioning the values for the `aws_vpc_id`. We use the provision defaults without being explicit in the brokerpak field.

Reply:
After [adding the skip_environment to azure.mssql_db_fog_server_pair_creds](https://github.com/pivotal/cloud-service-broker-tile-azure/commit/3774fa338aa5b53c49eca68b06be3a56ae27a545#diff-e1b86afc1785b5cd22896bcf751393d25a2cee50e7a26160ff9cd97158992395R498) the propagation you mention is not working.

I described the manual acceptance process I followed in [this Tracker Story](https://www.pivotaltracker.com/story/show/185362674/comments/237297211).

I don't know whether the `provisionDefaultOverrides` are computed using environment variables or there is something else going on.

I'm going to perform some tests with an older version without the `skip_environment` change to see if the propagation was working before.

Reply:
I believe this change is not simply for making propagation more explicit.
This change ties the default value to a property read from config.
According to [docs](https://github.com/cloudfoundry/cloud-service-broker/blob/f9d0df3465d3cb4d6cba98796998a178fc184682/cmd/config.go#L29-L38):
> The GCP Service Broker can be configured using both environment variables and
configuration files.
>
> It accepts configuration files in YAML, JSON, TOML and Java properties formats.
You can specify a configuration file to read using the --config argument.
>
> You can also specify configurations via environment variables.
The environment variables take the form GSB_<property> where property is the
same name as in the config file transformed to be in upper case and with all
dots replaced with underscores.

I could not find any reference in these docs telling that certain paths from the config
are used automatically used as defaults. However, the name of the config we are passing
`service.csb-azure-mssql-fog-run-failover.provision.defaults.server_pairs`
seems to suggest that we were expecting exactly that to happen.

Reply:
I ran some test [using version 1.4.1 of the tile](https://network.pivotal.io/products/cloud-service-broker-azure/#/releases/1239697) and the propagation doesn't work **as I would expect**.
Setting a value in `Azure SQL Failover Group Test Runner Config (csb-azure-mssql-fog-run-failover)` seems to be totally useless.
If I try to create a `csb-azure-mssql-fog-run-failover` instance it fails with:
> provision could not be completed: Service broker error: 3 error(s) occurred: (root): fog_instance_name is required; (root): server_pair_name is required; (root): server_pairs is required

Which makes absolutely no sense, since we have defined `server_pairs` in the tile.
Also, IMHO I believe this problem doesn't affect only consumers of the tile.
It also means that consumers of the brokerpak can't define `server_pairs` in the config file.

It MAY also mean that we are expecting a config entry with a path such as
`service.csb-azure-mssql-fog-run-failover.provision.defaults.server_pairs`
to be automatically mapped as a default value, when in fact this is not happening.

Reply:
A similar strategy is being used in other services as well:
https://github.com/cloudfoundry/csb-brokerpak-azure/blob/0a5ed5234b1d61ac7122f805bc00b81f3afcbe6c/azure-mssql-db.yml#L154

The only difference as I mentioned in the commit message is that the config we are dealing with
in this commit has a very misleading structure, suggesting that it will be used as default automatically
but it doesn't.

Reply:
I had a look at the order of resolution from csb docs and still can't find a different explanation than, it doesn't work as documented:
https://github.com/cloudfoundry/cloud-service-broker/blob/f9d0df3465d3cb4d6cba98796998a178fc184682/pkg/broker/service_definition.go#L370-L383

> variables function is used by ProvisionVariables and UpdateVariables
to get the variables and values to pass to Terraform. The output is a map where
the keys are all the variables that Terraform is expecting, and the values are their values.
Variables have a very specific resolution order. Lower number are overwritten by higher numbers
except where noted.

1. Global defaults: key/value pairs set in JSON in Viper `provision.defaults`
2. Service defaults: key/value pairs set in JSON in Viper `service.<service>.provision.defaults`
3. Request parameters: key/value pairs passed in the provision/update request, typically
    with the "-c" parameter of the "cf" command
4. Provision overrides: from the Plan definition "provision_overrides"
5. Default values: from the Service definition YAML "user_inputs", only if the value has not been set above.
6. Plan properties: from the Plan definition key "properties"
7. Computed values: from Service definition YAML "computed_inputs", only if the value has not been set above.

Reply:
This hypothesis could be explained by the fact that we are not using `Service defaults` functionality anywhere else.

So if this hypothesis holds true, it might be reasonable to get rid of such functionality.
After all, if no one has complained about this despite being presumably broken, it must mean no one is relying on this functionality.